ipmidetect/ipmidetectd.8.in
.\"#############################################################################
.\"$Id: ipmidetectd.8.in,v 1.11 2008-03-28 16:52:54 chu11 Exp $
.\"#############################################################################
.\"  Copyright (C) 2007-2008 Lawrence Livermore National Security, LLC.
.\"  Copyright (C) 2007 The Regents of the University of California.
.\"  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
.\"  Written by Albert Chu <chu11@llnl.gov>
.\"  UCRL-CODE-228523
.\"
.\"  This file is part of Ipmidetect, tools and libraries for detecting
.\"  IPMI nodes in a cluster. For details, see http://www.llnl.gov/linux/.
.\"
.\"  Ipmidetect is free software; you can redistribute it and/or modify it under
.\"  the terms of the GNU General Public License as published by the Free
.\"  Software Foundation; either version 2 of the License, or (at your option)
.\"  any later version.
.\"
.\"  Ipmidetect is distributed in the hope that it will be useful, but WITHOUT
.\"  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\"  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\"  for more details.
.\"
.\"  You should have received a copy of the GNU General Public License along
.\"  with Ipmidetect.  If not, see <http://www.gnu.org/licenses/>.
.\"#############################################################################
.TH ipmidetectd 8 "February 2007" "ipmidetectd @VERSION@" ipmidetectd
.SH "NAME"
ipmidetectd \- IPMI node detection monitoring daemon
.SH "SYNOPSIS"
.B ipmidetectd [OPTIONS]
.br
.SH "DESCRIPTION"
The
.B ipmidetectd
daemon regularly sends out IPMI pings to determine if nodes have
functioning IPMI interfaces or not.  It is primarily used to determine
those nodes in the cluster that are not functioning properly, allowing
IPMI commands to be executed across the cluster faster by skipping those
nodes that would automatically timeout.
.LP
The 
.B ipmidetectd
daemon must be configured using the 
.B ipmidetectd 
configuration file.  Until
hosts to be pinged are configured, 
.B ipmidetectd 
will not function properly.
.LP
See the
.BR ipmidetectd.conf(5)
manpage for additional information.

.SH "OPTIONS"
.TP
.B  -h, --help
Output help
.TP
.B  -v, --version
Output version
.if @WITH_DEBUG@ \{
.TP
.B  -c, --config-file
Specify alternate configuration file
.TP
.B  -d, --debug
Turn on debugging and run daemon in foreground
\}

.SH "FILES"
@IPMIDETECTD_CONFIG_FILE_DEFAULT@
.SH "ORIGIN"
Developed by Albert Chu <chu11@llnl.gov> on LLNL's Linux clusters.
.SH "SEE ALSO"
ipmidetectd.conf(5)


NEWS
0.6.1 - 00/00/00
----------------
For Users
---------
o Add new tool ipmi-sensors-config.
o Add new tool ipmi-oem.
o Support --groups (plural) for multiple groups listing in
  ipmi-sensors.  Older --group (non-plural) option maintained for
  backwards compatability.
o Support --list-groups and --flush in ipmimonitoring for consistency
  to ipmi-sensors.
o Change --cache-dir to --sdr-cache-dir in ipmimonitoring for consistency
  to ipmi-sensors.
o Change short option -f option to -n in pef-config and bmc-config for
  consistency to ipmi-sensors-config.  The -f option is maintained for
  backwards compatability.
o Support -e (--escape-char) for alternate escape character in
  ipmiconsole.
o Enhanced speed of ipmi-fru, by skipping unnecessary FRU data reads.
o SSIF Inband driver will timeout after a certain period of time.
o Support additional sensor monitoring interpretations in ipmimonitoring.
o Provide more detailed error messages in config tools.
o Support shared SDR caches for ipmi-sensors, ipmi-fru, ipmi-sel,
  ipmi-sensors-config, and ipmimonitoring.  (libipmimonitoring will
  still use a different default cache directory.)  This will require a
  new cache generation for these tools.
o Modified SDR caching output messages in ipmi-sensors, ipmi-sel,
  ipmi-fru, ipmi-sensors-config and ipmimonitoring to provide more
  useful information.
o In ipmi-sensors, updated SDR output under verbose and very verbose
  output modes with:
  o More IPMI compliant field names.  For example, 'Sensor Name' has
    been replaced with 'ID String'.
  o More correct output.  For example, 'NA' will be output instead of
    '0.0' when a value is not truly available.  Fields that don't exist
    will not have a default output.
  o Fix incorrect hysterisis calculation and output.
  o Fix incorrect OEM data output.
  o Support more SDR record types for output.
  o Output more SDR record information in general.
o Various other bug fixes, documentation fixes, and enhancements.

For Developers
--------------
o Many libfreeipmi API changes.  Not counting specific tool changes
  listed above, these API changes should not affect FreeIPMI user tools.
  o locate/probing API revamped.
  o SMIC driver/api removed. 
  o 'udm' renamed to 'api' globally.
  o 'ipmi_device_t' device context renamed to 'ipmi_ctx_t'.
  o New sdr-caching library added. 
  o Code re-organized into subdirectories.
  o New header files for more ipmi spec subsections added.
o Support new 'ipmi_monitoring_sdr_cache_filenames' function in
  libipmimonitoring.
o Store SDR with different format for ipmi-sensors, ipmi-fru,
  ipmi-sel, and ipmi-sensors-config, ipmimonitoring/libipmimonitoring
  for easier future extensibility.  
o Packet names added to IPMI dump outputs in all tools for easier
  packet identification in debugging.
o Support driver timeout error codes in all drivers and APIs.
o SDR record dumping added into all tools.
o Various other bug fixes and enhancements.

0.5.6 - 03/27/08
----------------
o Fix bmc-watchdog probing corner case.
o Fix bmc-watchdog openipmi corner case.

0.5.5 - 02/25/08
----------------
o Fix bmc-info firmware and IPMI version output.
o Fix Sun endian bug workaround.
o Fix FRU data output corner cases.
o Fix workaround documentation.

0.5.4 - 02/07/08
----------------
o Output addresses in hex in ipmi-locate.
o Clarify device addresses in manpages.
o Support --ignore-sdr-cache option in ipmi-sel and ipmi-fru.
o Minor SDR caching bug fixes.
o Minor ipmimonitoring corner case fixes.
o Various SSIF corner case fixes.

0.5.3 - 01/02/08
----------------
o Fix minor ipmimonitoring command line option combination bugs.
o Fix several minor corner cases in ipmi-sel, ipmi-sensors, and
ipmimonitoring.
o Fix corner case in all tools by recognizing 127.0.0.1 as the same as
localhost.
o Fix documentation/manpage text/typos.

0.5.2 - 12/21/07
----------------
o Clarify "Password Verification Timeout" troubleshooting information.
o Fix ipmipower cmdline parse bug.
o Support suffixes in hostrange inputs.

0.5.1 - 12/07/07
----------------
For Users
---------
o Added previously missing IPMI 2.0 (-D LAN_2_0) support into
  ipmi-chassis, ipmi-fru, ipmi-sensors, ipmi-sel, ipmi-raw, and
  ipmimonitoring.
o Added more strict IPMI over LAN implementation into ipmi-chassis,
  ipmi-fru, ipmi-sensors, ipmi-sel, ipmi-raw, and ipmimonitoring.  
o OpenIPMI and KCS Inband drivers will timeout after a certain period
  of time, so tools will no longer hang if the BMC is non-functional.
o Ported IPMI compliance workarounds from ipmipower, ipmiconsole, and
  ipmimonitoring into ipmi-chassis, ipmi-fru, ipmi-sensors, ipmi-sel,
  and ipmi-raw.
o Updated all manpages with more instructions, information, examples,
  and trouble shooting tips.
o Support --debug option under the non-debug builds.
o Rewrote error messages to provide more accurate descriptions.
o Removed CALLBACK and OEM privilege capabilities in most tools.
o Removed the quiet and silent options from bmc-config.
o Added openipmi driver to bmc-watchdog.
o Added keypair command line support to pef-config.
o Removed bmc-autoconfig tool.
o Added more details comments and instructions to bmc-config checkout.
o Converted ipmiconsole, ipmipower, and ipmimonitoring to use argp
  instead of getopt for consistency of usage output.
o Re-word "ASUS 2.0" workaround into generic "IGNORE SOL PAYLOAD SIZE"
  workaround.
o Support new "IGNORE SOL PORT" workaround.
o Made many command line options, interactive prompt, and config file
  options consistent across FreeIPMI tools.  Inconsistencies between
  dashes and underscores have been fixed in a number of places.
  Backwards compatability has been maintained when possible. Notable
  changes include:
  o --priv-level and --privilege options are now --privilege-level.
  o --auth-type options are now --authentication-type.
  o --reg-space or --register-space options are now --register-spacing.
  o --hostnames (plural) is now --hostname (not-plural).  

ipmidetect/src/ipmidetectd/ipmidetectd.c
/*****************************************************************************\
 *  $Id: ipmidetectd.c,v 1.7 2008-03-28 00:14:41 chu11 Exp $
 *****************************************************************************
 *  Copyright (C) 2007-2008 Lawrence Livermore National Security, LLC.
 *  Copyright (C) 2007 The Regents of the University of California.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Albert Chu <chu11@llnl.gov>
 *  UCRL-CODE-228523
 *
 *  This file is part of Ipmidetect, tools and libraries for detecting
 *  IPMI nodes in a cluster. For details, see http://www.llnl.gov/linux/.
 *
 *  Ipmidetect is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by the
 *  Free Software Foundation; either version 2 of the License, or (at your
 *  option) any later version.
 *
 *  Ipmidetect is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 *  for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with Ipmidetect.  If not, see <http://www.gnu.org/licenses/>.
\*****************************************************************************/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#if STDC_HEADERS
#include <string.h>
#endif /* STDC_HEADERS */
#include <syslog.h>
#include <sys/types.h>
#if HAVE_UNISTD_H
#include <unistd.h>
#endif /* HAVE_UNISTD_H */
#include <sys/stat.h>
#include <signal.h>
#include <assert.h>
#include <errno.h>

#include "ipmidetectd_config.h"
#include "ipmidetectd_loop.h"

#include "error.h"
#include "debug.h"

extern struct ipmidetectd_config conf;

static void
_daemon_init(void)
{
  /* Based on code in Unix network programming by R. Stevens */
  pid_t pid;
  int i;

  if ((pid = fork()) < 0)
    ERR_EXIT(("fork: %s", strerror(errno)));

  if (pid != 0)                 /* Terminate Parent */
    exit(0);

  setsid();

  if (signal(SIGHUP, SIG_IGN) == SIG_ERR)
    ERR_EXIT(("signal: %s", strerror(errno)));

  if ((pid = fork()) < 0)
    ERR_EXIT(("fork: %s", strerror(errno)));

  if (pid != 0)                 /* Terminate 1st Child */
    exit(0);

  chdir("/");

  umask(0);

  for (i = 0; i < 64; i++)
    close(i);
}

int
main(int argc, char **argv)
{
  err_init(argv[0]);
  err_set_flags(ERROR_STDOUT);

  ipmidetectd_config_setup(argc, argv);

#ifndef NDEBUG
  if (!conf.debug)
    {
      _daemon_init();
      err_set_flags(ERROR_SYSLOG);
    }
  else
    err_set_flags(ERROR_STDERR);
#else  /* NDEBUG */
  _daemon_init();
  err_set_flags(ERROR_SYSLOG);
#endif /* NDEBUG */

  /* Call after daemonization, since daemonization closes currently
   * open fds
   */
  openlog(argv[0], LOG_ODELAY | LOG_PID, LOG_DAEMON);

  ipmidetectd_loop();

  return 0;
}
